File: 19-removeNthFromEnd.php
<?php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {

    /**
     * @param ListNode $head
     * @param Integer $n
     * @return ListNode
     */
    function removeNthFromEnd($head, $n) {
        if ($head === null) return $head;
        $dummy = new ListNode(0);
        $dummy->next = $head;
        $slow = $dummy;
        $fast = $dummy;
        while ($n) {
            $fast = $fast->next;
            $n--;
        }
        while ($fast && $fast->next) {
            $slow = $slow->next;
            $fast = $fast->next;
        }
        $slow->next = $slow->next->next;
        return $dummy->next;
    }
}

File: go-code/19-removeNthFromEnd.go
package main

/**
删除
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
*/

// 删除链表的倒数第n个节点
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{}
	dummy.Next = head
	if head == nil {
		return head
	}
	fast := dummy
	slow := dummy
	for ; n > 0; n-- {
		fast = fast.Next
	}
	for fast != nil && fast.Next != nil {
		fast = fast.Next
		slow = slow.Next
	}
	slow.Next = slow.Next.Next
	return dummy.Next
}
